File: rest/src/main/java/io/sparkled/viewmodel/stage/StageViewModelConverterImpl.java
package io.sparkled.viewmodel.stage;

import io.sparkled.model.entity.Stage;
import io.sparkled.persistence.stage.StagePersistenceService;
import io.sparkled.viewmodel.exception.ViewModelConversionException;

import javax.inject.Inject;

public class StageViewModelConverterImpl extends StageViewModelConverter {

    private StagePersistenceService stagePersistenceService;

    @Inject
    public StageViewModelConverterImpl(StagePersistenceService stagePersistenceService) {
        this.stagePersistenceService = stagePersistenceService;
    }

    @Override
    public StageViewModel toViewModel(Stage model) {
        return new StageViewModel()
                .setId(model.getId())
                .setName(model.getName())
                .setWidth(model.getWidth())
                .setHeight(model.getHeight());
    }

    @Override
    public Stage toModel(StageViewModel viewModel) {
        final Integer stageId = viewModel.getId();
        Stage model = getStage(stageId);

        return model
                .setName(viewModel.getName())
                .setWidth(viewModel.getWidth())
                .setHeight(viewModel.getHeight());
    }

    private Stage getStage(Integer stageId) {
        if (stageId == null) {
            return new Stage();
        }

        return stagePersistenceService.getStageById(stageId)
                .orElseThrow(() -> new ViewModelConversionException("Stage with ID of '" + stageId + "' not found."));
    }
}

File: rest/src/main/kotlin/io/sparkled/viewmodel/stage/StageViewModelConverterImpl.kt
package io.sparkled.viewmodel.stage

import io.sparkled.model.entity.Stage
import io.sparkled.persistence.stage.StagePersistenceService
import io.sparkled.viewmodel.exception.ViewModelConversionException

import javax.inject.Inject

class StageViewModelConverterImpl @Inject
constructor(private val stagePersistenceService: StagePersistenceService) : StageViewModelConverter() {

    override fun toViewModel(model: Stage): StageViewModel {
        return StageViewModel()
                .setId(model.getId())
                .setName(model.getName())
                .setWidth(model.getWidth())
                .setHeight(model.getHeight())
    }

    override fun toModel(viewModel: StageViewModel): Stage {
        val stageId = viewModel.getId()
        val model = getStage(stageId)

        return model
                .setName(viewModel.getName())
                .setWidth(viewModel.getWidth())
                .setHeight(viewModel.getHeight())
    }

    private fun getStage(stageId: Int?): Stage {
        if (stageId == null) {
            return Stage()
        }

        return stagePersistenceService.getStageById(stageId)
                .orElseThrow { ViewModelConversionException("Stage with ID of '$stageId' not found.") }
    }
}
